ENIGMAsystem/SHELL/Collision_Systems/Polygon/Polygonfuncs.cpp

```diff
@@ -0,0 +1,688 @@
+        return -1;
+    const enigma::BoundingBox &box = inst1->$bbox_relative();
+    const double x1 = inst1->x, y1 = inst1->y,
+                 xscale1 = inst1->image_xscale, yscale1 = inst1->image_yscale,
+                 ia1 = inst1->image_angle;
+    int left1, top1, right1, bottom1;
+
+    get_border(&left1, &right1, &top1, &bottom1, box.left(), box.top(), box.right(), box.bottom(), x1, y1, xscale1, yscale1, ia1);
+
+    return fabs(hypot(min(left1 - x, right1 - x),
+                    min(top1 - y, bottom1 - y)));
+}
+
+double move_contact_object(int object, double angle, double max_dist, bool solid_only)
+{
+    enigma::object_collisions* const inst1 = ((enigma::object_collisions*)enigma::instance_event_iterator->inst);
+    if (inst1->sprite_index == -1 && (inst1->mask_index == -1)) {
+        return -4;
+    }
+
+    const double x = inst1->x, y = inst1->y;
+
+    if (collide_inst_inst(object, solid_only, true, x, y) != NULL) {
+        return 0;
+    }
+
+    const double DMIN = 1, DMAX = 1000; // Arbitrary max for non-positive values, 1000 fits with other implementations.
+    const double contact_distance = DMIN;
+    double mid_dist = max_dist; // mid_dist is used for the subtraction part.
+    if (max_dist <= 0) { // Use the arbitrary max for non-positive values.
+        max_dist = DMAX;
+    }
+    mid_dist = max_dist;
+    const double radang = (fmod(fmod(angle, 360) + 360, 360))*(M_PI/180.0);
+    const double sin_angle = sin(radang), cos_angle = cos(radang);
+
+    // Subtraction.
+
+    const int quad = int(angle/90.0);
+
+    const enigma::BoundingBox &box = inst1->$bbox_relative();
+    const double x1 = inst1->x, y1 = inst1->y,
+                 xscale1 = inst1->image_xscale, yscale1 = inst1->image_yscale,
+                 ia1 = inst1->image_angle;
+    int left1, top1, right1, bottom1;
+
+    get_border(&left1, &right1, &top1, &bottom1, box.left(), box.top(), box.right(), box.bottom(), x1, y1, xscale1, yscale1, ia1);
+
+    for (enigma::iterator it = enigma::fetch_inst_iter_by_int(object); it; ++it)
+    {
+        const enigma::object_collisions* inst2 = (enigma::object_collisions*)*it;
+        if (inst2->sprite_index == -1 && (inst2->mask_index == -1))
+            continue;
+        if (inst2->id == inst1->id || (solid_only && !inst2->solid))
+            continue;
+        const enigma::BoundingBox &box2 = inst2->$bbox_relative();
+        const double x2 = inst2->x, y2 = inst2->y,
+                     xscale2 = inst2->image_xscale, yscale2 = inst2->image_yscale,
+                     ia2 = inst2->image_angle;
+        int left2, top2, right2, bottom2;
+
+        get_border(&left2, &right2, &top2, &bottom2, box2.left(), box2.top(), box2.right(), box2.bottom(), x2, y2, xscale2, yscale2, ia2);
+
+        if (right2 >= left1 && bottom2 >= top1 && left2 <= right1 && top2 <= bottom1)
+        {
+            mid_dist = DMIN;
+            break;
+        }
+
+        switch (quad)
```

Comment:
For readability, it might be good to do one or more of these:
- Move the definition of `quad` closer to this switch() (it's only used here)
- Rename `quad` to  `quadrant`
- Inline the quadrant calculation and just comment "quadrant of the move_contact angle"

To be honest, I'm a little apprehensive of special-casing on the angle quadrant, as it's never worked out well for me, but your results are looking pretty good, so let's run with it. 🙂